expr/parse_test.go

```diff
@@ -635,6 +787,28 @@ func TestParse(t *testing.T) {
 				argsStr: "metric1 | func2(), 3",
 			},
 			nil,
+			"",
+		},
+		{
+			"func1(1 | func2(), 3)",
+			&expr{
+				str:   "func1",
+				etype: etFunc,
+				args: []*expr{
+					{
+						str:   "func2",
+						etype: etFunc,
+						args: []*expr{
+							{str: "1"},
```

Comment:
If we think of pipe syntax as just semantic sugar, and the "1" is effectively the first arg to func2, then shouldn't "1" be parsed as an int?

Reply:
I'm trying to match what graphite-web does as much as possible. 

graphite-web doesn't expect the input of a pipe to be a number. It can either be a template, call (i.e. function call), or a pathExpression (i.e. a metric name). https://github.com/graphite-project/graphite-web/blob/9e3e4e0aed8edf82004afc12aa6c55f83c5cbf01/webapp/graphite/render/grammar_unsafe.py#L138-L141

So it'll see the `1` as a _metric_ named "1" being piped into the function, rather than a number being piped into the function. 

---

Running `grammar.scanString("func1(1 | func2(2, 2a))")` in the graphite-web codebase and printing the output gives:

```
[[[0, 'func1', [['1', [[[10, 'func2', [['2']], [['2a', []]], 22]]]]], 23], []]]
- expression: [[0, 'func1', [['1', [[[10, 'func2', [['2']], [['2a', []]], 22]]]]], 23], []]
  - call: [0, 'func1', [['1', [[[10, 'func2', [['2']], [['2a', []]], 22]]]]], 23]
    - args: [[['1', [[[10, 'func2', [['2']], [['2a', []]], 22]]]]]]
      [0]:
        [['1', [[[10, 'func2', [['2']], [['2a', []]], 22]]]]]
        - expression: ['1', [[[10, 'func2', [['2']], [['2a', []]], 22]]]]
          - pathExpression: '1'
          - pipedCalls: [[[10, 'func2', [['2']], [['2a', []]], 22]]]
            - pipedCall: [[10, 'func2', [['2']], [['2a', []]], 22]]
              - call: [10, 'func2', [['2']], [['2a', []]], 22]
                - args: [[['2']], [['2a', []]]]
                  [0]:
                    [['2']]
                    - number: ['2']
                      - integer: '2'
                  [1]:
                    [['2a', []]]
                    - expression: ['2a', []]
                      - pathExpression: '2a'
                      - pipedCalls: []
                - end: 22
                - funcname: 'func2'
                - start: 10
    - end: 23
    - funcname: 'func1'
    - start: 0
  - pipedCalls: []
```

Reply:
I see. doing what graphite does is sensible.